### ISSUE TITLE
Fix/implement support for parallel recovery

### ISSUE THREAD

**Issue by jessecarterMOOSE** [idaholab/moose#8410]
### Description of the enhancement or error report
Multiple assertions occur when running the test suite in parallel and with recover.

Some of these assertions include:
- `mesh.comm().verify(mesh.n_elem()), libmesh/src/mesh/mesh_communication.C, line 980`
- `mesh.comm().semiverify(p_my_neighbor), libmesh/src/mesh/mesh_tools.C, line 1767`
- and others...

First assert seems to involve initial adaptivity upon recovering. A simple of example of a test failing:
https://github.com/idaholab/moose/blob/devel/test/tests/postprocessors/num_adaptivity_cycles/num_adaptivity_cycles.i
The error occurs almost immediately when running the recover portion of the test. The "Framework Information" doesn't even get printed out.

The second assert seems to involve parallel meshes. 

The remaining failures are for various reasons.

Running with >= 4 cores/test gives some Exodiffs and CSVdiffs too.

### Rationale for the enhancement or information for reproducing the error
Just run the test suite:
`./run_tests -j 24 -p 4 --recover --dbg`

Note only a subset of these will fail with:
`./run_tests -j 24 -p 2 --recover --dbg`

And all tests pass when all tests are run in serial with recovery.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

Less bugs.

**Comment by jessecarterMOOSE** [comment]
Hmm, the test linked above works if `Adaptivity/initial_steps` is set to 0 or 1.

**Comment by jessecarterMOOSE** [comment]
And these two tests that fail with the same assert as above:

- https://github.com/idaholab/moose/blob/devel/test/tests/markers/box_marker/box_marker_adapt_test.i
- https://github.com/idaholab/moose/blob/devel/test/tests/markers/error_tolerance_marker/error_tolerance_marker_adapt_test.i

are steady-state tests and shouldn't be run with --recover. However, the assert gets hit before that fact is discovered.

**Comment by jessecarterMOOSE** [comment]
BTW when you run with more cores per test, some Exodiff's and CSVdiff's start to show up too.

**Comment by jessecarterMOOSE** [comment]
OK, I'm trying to reverse engineer this thing, as always. Here's a basic test:

[test_recover_adaptivity_assert.i.txt](https://github.com/idaholab/moose/files/734195/test_recover_adaptivity_assert.i.txt)

Run normally, it produces this marker/mesh:
![baseline](https://cloud.githubusercontent.com/assets/12945054/22357883/2747fdfe-e409-11e6-90db-a1c9638e1367.png)

But then I run like this:
```
> rm -rf *out*
> mpiexec -np 2 ../../../moose_test-dbg -i test_recover_adaptivity_assert.i --half-transient Outputs/checkpoint=true
> mpiexec -np 2 ../../../moose_test-dbg -i test_recover_adaptivity_assert.i --recover
```
And, upon recovering, I get the error:
```
Assertion `mesh.comm().verify(mesh.n_elem())' failed.

Assertion `mesh.comm().verify(mesh.n_elem())' failed.

[1] ../src/mesh/mesh_communication.C, line 980, compiled Dec 16 2016 at 19:10:30
[0] ../src/mesh/mesh_communication.C, line 980, compiled Dec 16 2016 at 19:10:30
```

**Comment by jessecarterMOOSE** [comment]
Now, if I tweak the previous example ever-so-slightly, it _WILL_ run:

Set `max_h_level=1`:
![max-h-level-1](https://cloud.githubusercontent.com/assets/12945054/22358019/f254da76-e409-11e6-8aeb-ff566e432d5a.png)

Or just move the box:
![move-box](https://cloud.githubusercontent.com/assets/12945054/22358024/f9f6c442-e409-11e6-982f-29d4055a0ef4.png)

**Comment by jessecarterMOOSE** [comment]
It seems like it has something to do with that "extra" adaptivity (presumably to satisfy the incompatibilities of the different h-levels, but what the heck do I know) residing on the edge of the mesh?

Remember, the baseline test case, _DOES_ run in parallel, and _DOES_ run with ```--half-transient/--recover```, but it _DOES NOT_ work when doing both.

**Comment by permcody** [comment]
Thanks Jesse, Honestly what you are doing is probably what I would be doing. This is all really helpful.

**Comment by jessecarterMOOSE** [comment]
Well I've done about all I can with this. I sure hope you can make sense of it!

**Comment by rwcarlsen** [comment]
I think I may have solved this (at least partly).  See libmesh/libmesh#1222 for details

**Comment by rwcarlsen** [comment]
Went from
5 timeouts, 7 diffs, 15 crashes...
to
1 timeout, 7 diffs, and 12 crashes

Running ``./run_tests -j 24 -p 4 --recover --dbg`` without this patch:

```
mesh_modifiers/mesh_extruder.tri_test_part1................................................. FAILED (TIMEOUT)
mesh_modifiers/mesh_extruder.angle_test_part1............................................... FAILED (TIMEOUT)
mesh_modifiers/subdomain_bounding_box.inside_part1.......................................... FAILED (TIMEOUT)
mesh_modifiers/subdomain_bounding_box.oriented_inside_part1................................. FAILED (TIMEOUT)
mesh_modifiers/subdomain_bounding_box.oriented_outside_part1................................ FAILED (TIMEOUT)
executioners/executioner.test_steady........................................................ FAILED (EXODIFF)
mesh/custom_partitioner.custom_linear_partitioner........................................... FAILED (EXODIFF)
utils/spline_interpolation.bicubic_spline................................................... FAILED (EXODIFF)
nodalkernels/constant_rate.test............................................................. FAILED (EXODIFF)
userobjects/setup_interface_count.ElementUserObject......................................... FAILED (CSVDIFF)
userobjects/setup_interface_count.SideUserObject............................................ FAILED (CSVDIFF)
userobjects/setup_interface_count.InternalSideUserObject.................................... FAILED (CSVDIFF)
auxkernels/solution_aux.aux_nonlinear_solution_adapt_from_xda................................. FAILED (CRASH)
ics/function_scalar_ic.test_part1............................................................. FAILED (CRASH)
mesh/ghost_functors.geometric_edge_neighbor_one_2D............................................ FAILED (CRASH)
mesh/ghost_functors.geometric_edge_neighbor_two_2D............................................ FAILED (CRASH)
mesh/ghost_functors.geometric_edge_neighbor_two_3D_Mac........................................ FAILED (CRASH)
mesh/ghost_functors.geometric_edge_neighbor_one_3D_Mac........................................ FAILED (CRASH)
transfers/multiapp_mesh_function_transfer.fromsub_source_displaced_part1...................... FAILED (CRASH)
mesh/stitched_mesh.test....................................................................... FAILED (CRASH)
postprocessors/num_adaptivity_cycles.test..................................................... FAILED (CRASH)
postprocessors/num_dofs.test.................................................................. FAILED (CRASH)
materials/stateful_prop.spatial_adaptivity.................................................... FAILED (CRASH)
materials/stateful_prop.adaptivity............................................................ FAILED (CRASH)
mesh/unique_ids.parallel_mesh................................................................. FAILED (CRASH)
utils/random.test_par_mesh.................................................................... FAILED (CRASH)
utils/random.test_uo_par_mesh................................................................. FAILED (CRASH)
```

With this patch we get:

```
misc/line_source.test_part1................................................................. FAILED (TIMEOUT)
executioners/executioner.test_steady........................................................ FAILED (EXODIFF)
mesh/custom_partitioner.custom_linear_partitioner........................................... FAILED (EXODIFF)
nodalkernels/constant_rate.test............................................................. FAILED (EXODIFF)
utils/spline_interpolation.bicubic_spline................................................... FAILED (EXODIFF)
userobjects/setup_interface_count.ElementUserObject......................................... FAILED (CSVDIFF)
userobjects/setup_interface_count.SideUserObject............................................ FAILED (CSVDIFF)
userobjects/setup_interface_count.InternalSideUserObject.................................... FAILED (CSVDIFF)
ics/function_scalar_ic.test_part1............................................................. FAILED (CRASH)
mesh/ghost_functors.geometric_edge_neighbor_one_2D............................................ FAILED (CRASH)
mesh/ghost_functors.geometric_edge_neighbor_two_2D............................................ FAILED (CRASH)
mesh/ghost_functors.geometric_edge_neighbor_one_3D_Mac........................................ FAILED (CRASH)
mesh/ghost_functors.geometric_edge_neighbor_two_3D_Mac........................................ FAILED (CRASH)
mesh/stitched_mesh.test....................................................................... FAILED (CRASH)
transfers/multiapp_mesh_function_transfer.fromsub_source_displaced_part1...................... FAILED (CRASH)
materials/stateful_prop.adaptivity............................................................ FAILED (CRASH)
materials/stateful_prop.spatial_adaptivity.................................................... FAILED (CRASH)
mesh/unique_ids.parallel_mesh................................................................. FAILED (CRASH)
utils/random.test_par_mesh.................................................................... FAILED (CRASH)
utils/random.test_uo_par_mesh................................................................. FAILED (CRASH)
```

**Comment by jessecarterMOOSE** [comment]
Progress! :+1:

**Comment by jessecarterMOOSE** [comment]
The num_adaptivity_cycles test was the one I based my findings on above. That appears to pass now.

**Comment by roystgnr** [comment]
This appears to be a flaw in CheckpointIO: we're restoring prepartitioned mesh elements and nodes, but *without* restoring the remote_elem neighbor links.  Presumably @friedmud was assuming that our find_neighbors() method could be used to actually find all neighbors, but in reality that method name exists for backwards compatibility and brevity; the current method would be more accurately called find_neighbors_except_for_remote_elem() because a real find_neighbors() would require difficult and expensive communication.  libMesh relies on the mesh distribution code (e.g. MeshCommunication::gather_neighbors() or delete_remote_elements()) to get the remote_elem links right.

Fixing this properly (i.e. without gather_neighbors(), which is also expensive on large processor counts) will require an update to the CheckpointIO format.  Hopefully I can get that done this week.

**Comment by friedmud** [comment]
@roystgnr any word on this?  Our plan was to write out the remote_elem neighbors to the CheckpointIO file, right?

**Comment by jessecarterMOOSE** [comment]
:tada:

**Comment by roystgnr** [comment]
Sorry for the false positive, but I accidentally worded #8842 incorrectly; I didn't mean for it to autoclose this.

I do hope to have a patch in libMesh shortly which may be sufficient, though.

**Comment by roystgnr** [comment]
Okay; libMesh/libmesh#1291 was actually written to address other issues, but there's a chance it fixed this too.  I'll give it a try with the above listed test cases just as soon as I'm finished making sure I didn't regress anything else in the process.

**Comment by roystgnr** [comment]
With ```./run_tests -j 24 -p 4 --recover --dbg``` and libMesh/libmesh#1291 I see 13 diffs and 4 crashes:

```
nodalkernels/constant_rate.test............................................................. FAILED (EXODIFF)
time_steppers/postprocessor_dt.test......................................................... FAILED (EXODIFF)
time_steppers/constant_dt.constant_dt....................................................... FAILED (EXODIFF)
mesh/named_entities.test_periodic_names..................................................... FAILED (EXODIFF)
postprocessors/nodal_var_value.test_nodal_aux_var_value..................................... FAILED (EXODIFF)
postprocessors/mms_sine.test3dMesh.......................................................... FAILED (EXODIFF)
postprocessors/relative_solution_difference_norm.test....................................... FAILED (EXODIFF)
postprocessors/old_older_values.test_old.................................................... FAILED (EXODIFF)
postprocessors/old_older_values.test_older.................................................. FAILED (EXODIFF)
mesh/custom_partitioner.custom_linear_partitioner........................................... FAILED (EXODIFF)
userobjects/setup_interface_count.ElementUserObject......................................... FAILED (CSVDIFF)
userobjects/setup_interface_count.SideUserObject............................................ FAILED (CSVDIFF)
userobjects/setup_interface_count.InternalSideUserObject.................................... FAILED (CSVDIFF)
materials/stateful_prop.spatial_adaptivity.................................................... FAILED (CRASH)
materials/stateful_prop.adaptivity............................................................ FAILED (CRASH)
transfers/multiapp_mesh_function_transfer.fromsub_source_displaced_part1...................... FAILED (CRASH)
mesh/stitched_mesh.test....................................................................... FAILED (CRASH)
```

So we're not fixed, but more progress at least?

**Comment by permcody** [comment]
You can't necessarily draw the conclusion that this is your issue. Some of
these tests might diff with 4 processors without recover or there might be
problems unrelated to your patches. Do they diff with these same option on
2 processors (where we normally test)?

**Comment by roystgnr** [comment]
I'll check out the diffs now.

But there's definitely still some issue IMHO.  One of those crashes triggers on 2 processors, and it shows the same symptoms of something that I've been able to trigger in other tests on 8 processors.

**Comment by roystgnr** [comment]
Okay; using both libMesh/libmesh#1291 and #8853 gets me down to 2 crashes, 7 diffs, and 2 timeouts (shouldn't have been running other things on the same box...)

```
bcs/periodic.testperiodic_dp................................................................ FAILED (TIMEOUT)
executioners/transient_sync_time.testsynctimes_part1........................................ FAILED (TIMEOUT)
interfacekernels/1d_interface.test.......................................................... FAILED (EXODIFF)
mesh/custom_partitioner.custom_linear_partitioner........................................... FAILED (EXODIFF)
nodalkernels/constant_rate.threaded......................................................... FAILED (EXODIFF)
restart/restart.test_1....................................................................... FAILED (ERRMSG)
userobjects/setup_interface_count.ElementUserObject......................................... FAILED (CSVDIFF)
userobjects/setup_interface_count.SideUserObject............................................ FAILED (CSVDIFF)
userobjects/setup_interface_count.InternalSideUserObject.................................... FAILED (CSVDIFF)
transfers/multiapp_mesh_function_transfer.fromsub_source_displaced_part1...................... FAILED (CRASH)
mesh/stitched_mesh.test....................................................................... FAILED (CRASH)
```

**Comment by permcody** [comment]
This is exciting!
"Setup interface count"? That sounds like somebody's misguided use of a static or something just by the name alone! It looks a few of these should be easy to sort out.

**Comment by jessecarterMOOSE** [comment]
@roystgnr Did this get accidentally closed again?

**Comment by roystgnr** [comment]
Oh, hell, I need to get better at wording my PR messages.

**Comment by roystgnr** [comment]
Yeah, that libMesh commit fixed a bunch of tests, and the MOOSE PR I just put in should fix one more, but that leaves at least 8-10 tests still-failing.

What's worse, I'm going to have to drop this for a few weeks to catch up on work from other projects.

The good news is, as @permcody said, most of the remaining failures look suspiciously specific, so there's a chance that (like the StitchedMesh failure turned out to be) they're issues with just one feature, rather than corner cases which expose a core recovery functionality bug.  The obvious exception is restart.test_1, with its "Cannot restart using a different number of processors!" error; I'll try to at least glance at that and see if there's an obvious fix before I vanish on you guys.

**Comment by roystgnr** [comment]
The fixes on other tickets seem to be helping out here too.  With ```./run_tests -j 24 -p 4 --recover --dbg``` we're now down to only one test not passing for me, an assertion failure in the PointLocatorTree code in ```transfers/multiapp_mesh_function_transfer.fromsub_source_displaced_part1```.

My build settings differ a bit from the stuff on Civet, though.  Could @brianmoose run a sweep from 1 to 16 processors, using --recover this time, whenever it's convenient?

**Comment by brianmoose** [comment]
Sure, here it is: https://www.moosebuild.org/job/105866/

**Comment by roystgnr** [comment]
Only one failure (```postprocessors/scalar_coupled_postprocessor.test``` at 15 processors), and it's a known failure for that test without --recover.

I am a bit confused by Civet's success in the test that failed for me.  An overzealous or a non-critical assertion, possibly, so that the test ends up succeeding in opt mode?

We may be nearly in perfect shape with distributed recovery, too, once https://github.com/libMesh/libmesh/pull/1390 gets shined up and merged.  With ```./run_tests -p 4 --distributed-mesh --recover --dbg``` I get a parallel_node.C assertion failure in ```materials/output.block_displaced```, but everything else passes.  We'll want to do a big sweep with "--distributed-mesh --recover" in the MOOSE PR that ends up merging that libMesh PR.

Although... I may have an interesting bug to fix first.  It looks like the materials/output.block_displaced test is a serious failure case.  I'll start that in a separate ticket.

**Comment by brianmoose** [comment]
I know you probably aren't ready for it but I added the `--distributed-mesh --recover` sweep to the nightlies: https://www.moosebuild.org/job/106309/
Also, if you want any of these sweeps to be optional on libmesh PRs, just let me know. It is easy to do.

**Comment by roystgnr** [comment]
Definitely wasn't ready for it, as the plethora of crashes sadly confirmed.

#9633 might be ready for it, though, if you want to try a sweep there.

I'd love to add these as options on libMesh PRs!  I'd never actually noticed before that I had access to activate the optional tests.  I promise to only abuse this privilege a little bit.

**Comment by roystgnr** [comment]
I'm confused by the behavior of restart/new_dt.test with --recover turned on.  Is the normal test a recovery, and with --recover the test also does a recovery within each of the two recovery-separated halves of the normal test?

**Comment by permcody** [comment]
Head -> Explode. Mixing restart and recovery is definitely a lot to think about. The TestHarness tries to split every transient test unless recovery is disabled. So if both halves of the restart test are both transient, then yes, I suppose it's possible to have a four part test. Whether or not we care if that works is another story. I guess theoretically it should...

**Comment by roystgnr** [comment]
It gives the weirdest errors.  I try ```METHOD=dbg ./run_tests -p 1 --distributed-mesh --recover --recoversuffix cpa --re restart/new_dt.test``` and it fails with
```
restart/new_dt.test_part1_part1: *** ERROR ***
restart/new_dt.test_part1_part1: An output file with the name, new_dt_out_cp/0000, already exists.
restart/new_dt.test_part1_part1: 
restart/new_dt.test_part1_part1: 
restart/new_dt.test_part1_part1: Stack frames: 13
restart/new_dt.test_part1_part1: 0: libMesh::print_trace(std::ostream&)
restart/new_dt.test_part1_part1: 1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
restart/new_dt.test_part1_part1: 2: void mooseError<char const (&) [31], OutFileBase const&, char const (&) [18]>(char const (&) [31], OutFileBase const&, char const (&) [18])
restart/new_dt.test_part1_part1: 3: OutputWarehouse::addOutputFilename(OutFileBase const&)
```

I presume because both the restart file specified in the input file and the recovery file I specified from the test harness modifications have the same filename?  Hmm.. I can probably change the input file to use a section the way the test harness has to for cpa output, and then it'll just be a single output requirement.

That problem is unrelated to (although it's been making it harder to debug) the actual failure on 8 processors, though.  Civet just sees convergence failures in opt mode.  Running myself in dbg mode, there seems to be an I/O consistency assertion failing.  I'm not 100% sure it's a real error rather than a race condition from multiple uses of the same file, though.

**Comment by roystgnr** [comment]
Nah; I hate disabling failing tests unless I know *why* they're failing.  The assertion failure here looks like a bug that we had years ago, which was a disaster from a reliability perspective: basically 0.01% or some even tinier fraction of libMesh xdr/xda restart outputs were corrupted.  It seems unlikely that a new version of that bug would just *happen* to crop up in a weird corner case, but best to be sure.  I can replicate now with ASCII checkpoint files, so it should be easy enough to debug.

**Comment by roystgnr** [comment]
Actually, no, I can't yet replicate with ASCII... --recoversuffix cpa changes the recovery files to ASCII, but the restart files remain binary, and the restart file set is what's failing to load properly.

**Comment by roystgnr** [comment]
Okay, back to ASCII.

**Comment by roystgnr** [comment]
This looks like an actual bug, not just recovery files and restart files stepping on each other's toes.  Somehow node processor ids are getting reassigned in between the restart mesh read and restart solution read.

**Comment by roystgnr** [comment]
It looks like we're *not* disabling repartitioning or renumbering on restart meshes?  If that's the case then I'm just shocked it didn't bite us before.  Sure, *usually* you write out a mesh with a partitioning such that your partitioner is idempotent, but unless you guarantee that that's the case then you can't safely use a distributed restart file, only a serialized restart file.

**Comment by roystgnr** [comment]
Okay; disabling repartitioning and renumbering during restarts fixes the broken test, at least on 1 through 24 processors.  But...

1) We can only do that safely on parallel-aware mesh formats: just Nemesis and CheckpointIO, basically.  Nothing else encodes all the partitioning data we need to allow read() to skip a partitioner call entirely.

2) If we don't consider serial mesh formats then we have a scary inadequate fix, which would work *nearly* all the time but lead to crashes or corrupt data on infrequent occasions.  What happens when you're doing a restart on a GeneratedMesh, or an Exodus mesh, or basically anything other than Nemesis-or-CheckpointIO, and we tweak our METIS/ParMETIS settings in between when you saved a distributed restart and when you load it?

Hmm... digging into the code, it looks like XDA always writes serial restart files, and distributed restarts are only written simultaneously with CheckpointIO mesh files?  In that case all we'd need to do is make sure that, whenever we're doing a distributed restart, we load the restart mesh from file rather than using the existing mesh generation settings.  That should be doable.  I'm running other restart tests to confirm or refute my understanding of the problem, and if I'm right, then I'll try to get it into a PR ASAP.

Sorry for spamming everyone with my thinking out loud; I'm just not sure that "ASAP" will mean "before next week", and if it's not then I want notes to look back on.

**Comment by permcody** [comment]
Yes it should have, I probably should have CC'd you on the message I just sent to Jesse on the mailing list. I'll forward to you so you can see the discussion.

**Comment by roystgnr** [comment]
Gah, wrong button.  I'd really like to close this, but one of the two_circle_marker tests fails on the distributed recover tests, with a relative exodiff of like 0.1%, which is way too big to be a simple solver screw up.  We *also* fail on the plain distributed tests, but until I understand why I don't want to swear that there isn't a separate recovery problem.
